### ISSUE TITLE
Prysm Beacon Node Compatability

### ISSUE THREAD

**Issue by MysticRyuujin** [emeraldpay/dshackle#45]
I was wondering if you could look into why `prysm` cannot use `dshackle` as its ETH1 target.

In the `prysm` config I used:
`http-web3provider: "http://dshackle:8545/eth"`

For `dshackle` I'm simply pointing to infrua's goerli endpoint.

`dshackle` provides no errors or indication of failure and it is working just fine with Infura configured, but `prysm` throws a single error on startup:

```
time="2020-08-18 16:29:09" level=error msg="Could not connect to powchain endpoint" error="could not dial eth1 nodes: json: cannot unmarshal number into Go value of type string" prefix=powchain
```
Not sure if this is a prysm issue or a dshackle issue but I thought I would start here.

If I point `prysm` at the same infura endpoint, it works just fine.

**Comment by splix** [comment]
I guess "_cannot unmarshal number into Go value of type string_" means that some of JSON gives plain number where some hex encoded string is expected. May be some of the _default_ endpoints is providing such json, i.e. a method like `net_peerCount` or something.

Can you please catch the request for debugging? Like with Wireshark maybe?

**Comment by MysticRyuujin** [comment]
This may be related to #22 I talked to the Prysm guys and they said that the first call made is eth_chainId

I have two questions:

1. Was your fix to simply allow chainId to not be a string when Nethermimd returns it? If that's the case perhaps I simply need to take this up with the Nethermimd team.

2. Alternatively, should chainId be a static return directly from dshackle instead of asking the node (s) every time? I know some methods are handled by dshackle directly, wondering if this is or should be one of them.

**Comment by MysticRyuujin** [comment]
I looked at this again and I think dshackle isn't returning chainId

Nethermind Reply:
```
curl -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":20,"method":"eth_chainid","params":[]}' http://nethermind01:8545

{"jsonrpc":"2.0","result":"0x1","id":20}
```

dshackle doesn't return anything, just an HTTP 200, nor does it throw an error

I thought MAYBE I just needed to add `eth_chainId` to the methods enabled, but that didn't make any difference.

**Comment by MysticRyuujin** [comment]
Ok I take it back, it looks like the "methods enabled" is case sensitive and I needed to use "eth_chainId" not "eth_chainid"

I added:
```
methods:
        enabled:
          - name: eth_chainId
```

and now it's working...

So I think the fix here is to make eth_chainId a default supported call? Be it forwarded to the node(s) or answered directly by dshackle.

**Comment by MysticRyuujin** [comment]
Looking at the code, here: https://github.com/emeraldpay/dshackle/blob/fd070d5df47080beade492ab94aebf8b59ced795/src/main/kotlin/io/emeraldpay/dshackle/upstream/calls/DefaultEthereumMethods.kt#L129

It looks like you started to make eth_chainId a native/static call? but Goerli isn't listed. Should be "0x5"

**Comment by splix** [comment]
I see, that is something I was suspecting. Dshackle comes with networks which were in our roadmap, and Goerly wasn't in it. I see your PR, thank you, I think it would work, but I'll need to add another change to make it working

**Comment by splix** [comment]
Can you please try it with v0.8.2?

**Comment by MysticRyuujin** [comment]
I think it's getting closer :smile: 

But right now I'm getting this cryptic error from the beacon node, repeating over and over:

> Unable to process past logs could not query block with height 3276865: missing required field 'stateRoot' for Header

I've tried it with my own geth node, and with infura (I just swapped the `role: fallback` on them to test both)

```
time="2020-08-23 21:22:02" level=info msg="Starting beacon node" prefix=node version="Prysm/v1.0.0-alpha.22/0a5ec502b0207c0a39484fa29331fabb351858e7. Built at: 2020-08-18 02:18:25+00:00"
time="2020-08-23 21:22:02" level=info msg="RPC-API listening on port" address="0.0.0.0:4000" prefix=rpc
time="2020-08-23 21:22:02" level=warning msg="You are using an insecure gRPC server. If you are running your beacon node and validator on the same machines, you can ignore this message. If you want to know how to enable secure connections, see: https://docs.prylabs.network/docs/prysm-usage/secure-grpc" prefix=rpc
time="2020-08-23 21:22:02" level=info msg="Starting JSON-HTTP API" address="127.0.0.1:3500" prefix=gateway
time="2020-08-23 21:22:02" level=info msg="Connected to eth1 proof-of-work chain" endpoint="http://dshackle:8545/goerli" prefix=powchain
time="2020-08-23 21:22:03" level=error msg="Unable to process past logs could not query block with height 3276865: missing required field 'stateRoot' for Header" prefix=powchain
time="2020-08-23 21:22:08" level=info msg="New gRPC client connected to beacon node" addr="172.18.0.5:49718" prefix=rpc
time="2020-08-23 21:22:09" level=info msg="Connected to eth1 proof-of-work chain" endpoint="http://dshackle:8545/goerli" prefix=powchain
time="2020-08-23 21:22:09" level=error msg="Unable to process past logs could not query block with height 3276869: missing required field 'stateRoot' for Header" prefix=powchain
time="2020-08-23 21:22:13" level=info msg="New gRPC client connected to beacon node" addr="172.18.0.2:52160" prefix=rpc
time="2020-08-23 21:22:15" level=info msg="Connected to eth1 proof-of-work chain" endpoint="http://dshackle:8545/goerli" prefix=powchain
time="2020-08-23 21:22:15" level=error msg="Unable to process past logs could not query block with height 3276869: missing required field 'stateRoot' for Header" prefix=powchain
```

Here is my dshackle config:
```
Host: 0.0.0.0
port: 2449
tls:
  enabled: false

proxy:
  host: 0.0.0.0
  port: 8545
  tls:
    enabled: false
  routes:
    - id: goerli
      blockchain: goerli

cluster:
  defaults:
    - chains:
        - goerli
      options:
        min-peers: 10
  upstreams:
    - id: geth
      chain: goerli
      labels:
        provider: geth
        nodetype: full
      methods:
        enabled:
          - name: eth_chainId
      connection:
        ethereum:
          rpc:
            url: "http://geth:8545"
          ws:
            url: "ws://geth:8545"
    - id: infura
      chain: goerli
      role: fallback
      labels:
        provider: infura
        nodetype: full
      methods:
        enabled:
          - name: eth_chainId
      connection:
        ethereum:
          rpc:
            url: "https://goerli.infura.io/v3/<key>"
          ws:
            url: "wss://goerli.infura.io/ws/v3/<key>"
```

And here are the logs from dshackle:
```
2020-23-08 21:25:20.481 | INFO  |            StarterKt | Starting StarterKt on f855d25ed3ac with PID 1 (/app/classes started by root in /)
2020-23-08 21:25:20.488 | INFO  |            StarterKt | No active profile set, falling back to default profiles: default
2020-23-08 21:25:24.628 | INFO  |               Config | Using config: /etc/dshackle/dshackle.yaml
2020-23-08 21:25:25.433 | INFO  |           GrpcServer | Starting gRPC Server...
2020-23-08 21:25:25.435 | INFO  |           GrpcServer | Listening Native gRPC on 127.0.0.1:2449
2020-23-08 21:25:25.649 | WARN  |             TlsSetup | Using insecure transport for Native gRPC
2020-23-08 21:25:25.902 | INFO  |           GrpcServer | GRPC Server started
2020-23-08 21:25:25.930 | INFO  |          ProxyServer | Listening Proxy on 0.0.0.0:8545
2020-23-08 21:25:25.959 | WARN  |             TlsSetup | Using insecure transport for proxy
2020-23-08 21:25:26.184 | INFO  |  ConfiguredUpstreams | Using Goerli Testnet upstream, at http://geth:8545, ws://geth:8545
2020-23-08 21:25:26.193 | INFO  | WsFactory$EthereumWs | Connecting to WebSocket: ws://geth:8545
2020-23-08 21:25:26.552 | INFO  |  EthereumRpcUpstream | Configured for Goerli Testnet
2020-23-08 21:25:26.831 | INFO  | entMultistreamHolder | Upstream geth with chain TESTNET_GOERLI has been added
2020-23-08 21:25:26.832 | INFO  |  ConfiguredUpstreams | Using Goerli Testnet upstream, at https://goerli.infura.io/v3/bb8fc294c4b04189ad7f02525e524d90, wss://goerli.infura.io/ws/v3/bb8fc294c4b04189ad7f02525e524d90
2020-23-08 21:25:26.833 | INFO  | WsFactory$EthereumWs | Connecting to WebSocket: wss://goerli.infura.io/ws/v3/bb8fc294c4b04189ad7f02525e524d90
2020-23-08 21:25:27.796 | INFO  |  EthereumRpcUpstream | Configured for Goerli Testnet
2020-23-08 21:25:27.891 | INFO  | entMultistreamHolder | Upstream infura with chain TESTNET_GOERLI has been added
2020-23-08 21:25:29.080 | INFO  |            StarterKt | Started StarterKt in 9.633 seconds (JVM running for 11.701)
2020-23-08 21:25:30.091 | INFO  |          Multistream | State of GOERLI: height=?, status=UNAVAILABLE/2, lag=[0, 9223372036854775807]
2020-23-08 21:25:45.061 | INFO  |          Multistream | State of GOERLI: height=?, status=OK/1,SYNCING/1, lag=[0, 9223372036854775807]
2020-23-08 21:26:00.061 | INFO  |          Multistream | State of GOERLI: height=?, status=OK/1,SYNCING/1, lag=[0, 9223372036854775807]
2020-23-08 21:26:14.060 | INFO  |          Multistream | State of GOERLI: height=3277908, status=OK/2, lag=[0, 0]
```

**Comment by MysticRyuujin** [comment]
At first I thought it might be an `eth_getLogs` call but I added that as an available method but it didn't work. I asked the Prysm team and they think it's an `eth_call` for a range of blocks....

**Comment by MysticRyuujin** [comment]
Ok I think I've narrowed down the problem to the difference in these 2 RPC calls...

When calling `eth_getBlockByNumber` and specifying that the `boolean` be `true` (i.e. return full transaction object) the stateRoot is not returned as a property.
```sh
chase@RYUUJIN:~$ curl -s -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":7,"method":"eth_getBlockByNumber","params":["0x320d0c",true]}' http://validator01:8545/goerli | jq
{
  "jsonrpc": "2.0",
  "id": 7,
  "result": {
    "number": "0x320d0c",
    "hash": "0x97a05f720605d8c00836dd06085f60c0e2399aef3fc58ca89497566c5c56a0c9",
    "timestamp": "0x5f43a803",
    "transactions": [],
    "parentHash": "0x61fd1842706e4cf84bcc623c75f504c457e02f198b5cb68e26a8722427e3be4e",
    "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "miner": "0x0000000000000000000000000000000000000000",
    "difficulty": "0x1",
    "totalDifficulty": "0x490f1a",
    "size": 609,
    "gasLimit": "0x7a1200",
    "gasUsed": "0x0",
    "extraData": "0x476f65726c6920496e697469617469766520417574686f72697479000000000092667679eb736ba430874b683c7f0bc0a8076176628aec1bc25d99f51d4458b242a6e1976968e49c9e3bc5d1801aac04b3515e2b4bf568d9a4092124024104c101",
    "uncles": []
  }
}
```
However, if you set it to `false` (i.e. only return the hashes), the stateRoot is returned.
```sh
chase@RYUUJIN:~$ curl -s -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":7,"method":"eth_getBlockByNumber","params":["0x320d0c",false]}' http://validator01:8545/goerli | jq
{
  "jsonrpc": "2.0",
  "id": 7,
  "result": {
    "difficulty": "0x1",
    "extraData": "0x476f65726c6920496e697469617469766520417574686f72697479000000000092667679eb736ba430874b683c7f0bc0a8076176628aec1bc25d99f51d4458b242a6e1976968e49c9e3bc5d1801aac04b3515e2b4bf568d9a4092124024104c101",
    "gasLimit": "0x7a1200",
    "gasUsed": "0x0",
    "hash": "0x97a05f720605d8c00836dd06085f60c0e2399aef3fc58ca89497566c5c56a0c9",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "miner": "0x0000000000000000000000000000000000000000",
    "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "nonce": "0x0000000000000000",
    "number": "0x320d0c",
    "parentHash": "0x61fd1842706e4cf84bcc623c75f504c457e02f198b5cb68e26a8722427e3be4e",
    "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
    "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "size": "0x261",
    "stateRoot": "0x6e0c342b126fe393fa026c53655890538f49bd6dec71a5519880aaec3e9d2a08",
    "timestamp": "0x5f43a803",
    "totalDifficulty": "0x490f1a",
    "transactions": [],
    "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
    "uncles": []
  }
}
```

It almost seems to me that dshackle is doing the opposite thing it should do, however, when i query the nodes directly, either `true` or `false` it returns stateRoot...so I'm not sure what's going on

**Comment by splix** [comment]
Dshackle is actually always uses a block JSON that comes with `false` request. If the client makes request with `true`, it simply puts transactions into it, keeping everything else as is. 

I.e., if you make a request for a block with all transactions details (`true`), it splits it into block request and transactions requests internally, then merges into single response. Each of the elements may be cached individually, or otherwise requested in parallel from many upstreams. It's implemented for better caching.

So I expect it to always give `stateRoot` field. But the error log seems to say that the field is actually missing. That's strange.

**Comment by MysticRyuujin** [comment]
It is missing though, see the 1st query I made, when the Boolean is true, dshackle does not return it.

**Comment by splix** [comment]
Ah, I didn't realize it's actual response from Dshackle. What if you clear the Redis cache?

**Comment by splix** [comment]
It's very strange that it gives completely different JSON. Because it shouldn't modify it all, it processes it on a byte level.

**Comment by MysticRyuujin** [comment]
I can tell you with 100% certainty that the response is different :)

I am running geth and dshackle in docker containers. geth is exposed on port 8546 and dshackle is exposed at port 8545.

Proof:

```
chase@validator01:~$ docker container list
CONTAINER ID        IMAGE                                                     COMMAND                  CREATED             STATUS              PORTS                                                                                       NAMES
5d65fd3ca153        ethereum/client-go:v1.9.19                                "geth --datadir=/cha…"   2 minutes ago       Up 2 minutes        8546-8547/tcp, 0.0.0.0:30303->30303/tcp, 0.0.0.0:30303->30303/udp, 0.0.0.0:8546->8545/tcp   geth
5670158048fb        eth2validator_dshackle                                    "java -Xms1024m -cp …"   3 minutes ago       Up 3 minutes        0.0.0.0:2449->2449/tcp, 0.0.0.0:8545->8545/tcp, 8090/tcp                                    dshackle
```

In case you're wondering I'm building dshackle locally with a Dockerfile that is simply this:
```
FROM emeraldpay/dshackle:0.8.2
COPY dshackle.yaml /etc/dshackle/dshackle.yaml
```

dshackle is now configured ONLY to talk to my local geth node.

Proof:
```
chase@validator01:~$ docker logs --tail 100 dshackle
                                _     _     __ _     _                _    _
                               | |   | |   / /| |   | |              | |  | |
   ___ _ __ ___   ___ _ __ __ _| | __| |  / /_| |___| |__   __ _  ___| | _| | ___
  / _ \ '_ ` _ \ / _ \ '__/ _` | |/ _` | / / _` / __| '_ \ / _` |/ __| |/ / |/ _ \
 |  __/ | | | | |  __/ | | (_| | | (_| |/ / (_| \__ \ | | | (_| | (__|   <| |  __/
  \___|_| |_| |_|\___|_|  \__,_|_|\__,_/_/ \__,_|___/_| |_|\__,_|\___|_|\_\_|\___|
   Emerald Dshackle - Fault Tolerant Load Balancer for Blockchain API
   https://github.com/emeraldpay/dshackle
   v0.8.2 built from 08b59f6511 on 2020-08-24T01:56:59 UTC

2020-24-08 16:17:24.811 | INFO  |            StarterKt | Starting StarterKt on 5670158048fb with PID 1 (/app/classes started by root in /)
2020-24-08 16:17:24.820 | INFO  |            StarterKt | No active profile set, falling back to default profiles: default
2020-24-08 16:17:30.064 | INFO  |               Config | Using config: /etc/dshackle/dshackle.yaml
2020-24-08 16:17:30.962 | INFO  |           GrpcServer | Starting gRPC Server...
2020-24-08 16:17:30.962 | INFO  |           GrpcServer | Listening Native gRPC on 127.0.0.1:2449
2020-24-08 16:17:31.301 | WARN  |             TlsSetup | Using insecure transport for Native gRPC
2020-24-08 16:17:31.709 | INFO  |           GrpcServer | GRPC Server started
2020-24-08 16:17:31.776 | INFO  |          ProxyServer | Listening Proxy on 0.0.0.0:8545
2020-24-08 16:17:31.839 | WARN  |             TlsSetup | Using insecure transport for proxy
2020-24-08 16:17:32.448 | INFO  |  ConfiguredUpstreams | Using Goerli Testnet upstream, at http://geth:8545, ws://geth:8545
2020-24-08 16:17:32.459 | INFO  | WsFactory$EthereumWs | Connecting to WebSocket: ws://geth:8545
2020-24-08 16:17:32.618 | ERROR | WsFactory$EthereumWs | Failed to connect to websocket at ws://geth:8545. Error: Failed to connect to ws://geth:8545
2020-24-08 16:17:32.771 | INFO  |  EthereumRpcUpstream | Configured for Goerli Testnet
2020-24-08 16:17:33.020 | INFO  | entMultistreamHolder | Upstream geth with chain TESTNET_GOERLI has been added
2020-24-08 16:17:34.194 | INFO  |            StarterKt | Started StarterKt in 10.412 seconds (JVM running for 12.724)
2020-24-08 16:17:35.208 | INFO  |          Multistream | State of GOERLI: height=?, status=UNAVAILABLE/1, lag=[0]
2020-24-08 16:17:50.173 | INFO  |          Multistream | State of GOERLI: height=?, status=UNAVAILABLE/1, lag=[0]
2020-24-08 16:18:05.174 | INFO  |          Multistream | State of GOERLI: height=?, status=OK/1, lag=[0]
2020-24-08 16:18:20.174 | INFO  |          Multistream | State of GOERLI: height=?, status=OK/1, lag=[0]
2020-24-08 16:18:34.173 | INFO  |          Multistream | State of GOERLI: height=3282437, status=OK/1, lag=[0]
```

Here is the call to dshackle:
```
chase@validator01:~$ curl -s -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":7,"method":"eth_getBlockByNumber","params":["0x320d0c",true]}' http://localhost:8545/goerli | jq
{
  "jsonrpc": "2.0",
  "id": 7,
  "result": {
    "number": "0x320d0c",
    "hash": "0x97a05f720605d8c00836dd06085f60c0e2399aef3fc58ca89497566c5c56a0c9",
    "timestamp": "0x5f43a803",
    "transactions": [],
    "parentHash": "0x61fd1842706e4cf84bcc623c75f504c457e02f198b5cb68e26a8722427e3be4e",
    "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "miner": "0x0000000000000000000000000000000000000000",
    "difficulty": "0x1",
    "totalDifficulty": "0x490f1a",
    "size": 609,
    "gasLimit": "0x7a1200",
    "gasUsed": "0x0",
    "extraData": "0x476f65726c6920496e697469617469766520417574686f72697479000000000092667679eb736ba430874b683c7f0bc0a8076176628aec1bc25d99f51d4458b242a6e1976968e49c9e3bc5d1801aac04b3515e2b4bf568d9a4092124024104c101",
    "uncles": []
  }
}
```

And here is the exact same call directly to geth:
```
chase@validator01:~$ curl -s -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":7,"method":"eth_getBlockByNumber","params":["0x320d0c",true]}' http://localhost:8546 | jq
{
  "jsonrpc": "2.0",
  "id": 7,
  "result": {
    "difficulty": "0x1",
    "extraData": "0x476f65726c6920496e697469617469766520417574686f72697479000000000092667679eb736ba430874b683c7f0bc0a8076176628aec1bc25d99f51d4458b242a6e1976968e49c9e3bc5d1801aac04b3515e2b4bf568d9a4092124024104c101",
    "gasLimit": "0x7a1200",
    "gasUsed": "0x0",
    "hash": "0x97a05f720605d8c00836dd06085f60c0e2399aef3fc58ca89497566c5c56a0c9",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "miner": "0x0000000000000000000000000000000000000000",
    "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "nonce": "0x0000000000000000",
    "number": "0x320d0c",
    "parentHash": "0x61fd1842706e4cf84bcc623c75f504c457e02f198b5cb68e26a8722427e3be4e",
    "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
    "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "size": "0x261",
    "stateRoot": "0x6e0c342b126fe393fa026c53655890538f49bd6dec71a5519880aaec3e9d2a08",
    "timestamp": "0x5f43a803",
    "totalDifficulty": "0x490f1a",
    "transactions": [],
    "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
    "uncles": []
  }
}
```

**Comment by splix** [comment]
That's definitely a bug then, somewhere it replaces the original bytes with an internal representation

**Comment by MysticRyuujin** [comment]
FYI: I have a suspicion that this issue is breaking ganche/truffle users

**Comment by MysticRyuujin** [comment]
An update:

Prysm has changed the way that they query the ETH1 node, and because of that it looks like `dshackle` is now compatible:
```
time="2020-09-22 10:31:46" level=info msg="Connected to eth1 proof-of-work chain" endpoint="http://dshackle:8545/goerli" prefix=powchain
```

I'm not sure if you would like me to close this issue and open a new issue that is specific to the issue above (incorrect response data for `eth_getBlockByNumber` when bool is `true`) or just keep tracking it here?

Either way I just thought I'd post an update in case anyone else was tracking.

**Comment by splix** [comment]
Thank for the update and reminding, almost forgot about it. I think I'll try to fix the bug with json then. This issue should be fine.
